backend/rates/coingecko.go

```diff
@@ -0,0 +1,153 @@
+
+		start := updater.histLatestTimestamp("BTC")
+		if !start.IsZero() {
+			// Start slightly past the last fetched timestamp
+			start = start.Add(10 * time.Minute)
+			now := time.Now()
+			// It's ok if start == now or now < start. CoinGecko simply returns no results.
+			// It's also ok if our "now" doesn't match CoinGecko: it always returns
+			// the latest available data without errors.
+			err := updater.updateHist("BTC", "USD", start, now)
+			if err != nil {
+				updater.log.Printf("updateHist(%v, %v): %v", start, now, err)
+				untilNext = time.Minute // TODO: exponential backoff
+			}
+		}
+
+		time.Sleep(untilNext)
+	}
+}
+
+// backwardHistoryLoop updates historical market exchange rates starting with
+// the earliest fetched timestamp backwards in time.
+func (updater *RateUpdater) backwardHistoryLoop() {
+	for {
+		// When to update next, after this loop iteration is done.
+		untilNext := 10 * time.Minute // TODO: add jitter
+
+		end := updater.histEarliestTimestamp("BTC")
+		if end.IsZero() {
+			end = time.Now()
+		}
+		// 90 days is the max interval which CoinGecko responds with hourly timeseries.
+		start := end.Add(-90 * 24 * time.Hour)
+		if time.Since(start) > 720*24*time.Hour {
+			// ~2 years of historical data is enough for now.
+			// TODO: Where do we stop?
+			return
+		}
+
+		err := updater.updateHist("BTC", "USD", start, end)
+		if err != nil {
+			updater.log.Printf("updateHist(%s, %s): %v", start, end, err)
+			untilNext = time.Minute // TODO: exponential backoff
+		}
+
+		time.Sleep(untilNext)
+	}
+}
+
+func (updater *RateUpdater) updateHist(coin, fiat string, start, end time.Time) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+	rates, err := updater.fetchMarketRange(ctx, coin, fiat, start, end)
+	if err != nil {
+		return err
+	}
+
+	updater.histMu.Lock()
+	defer updater.histMu.Unlock()
+	// TODO: updater.hist[coin] and rates may have duplicate timestamps.
+	a := append(updater.hist[coin], rates...)
+	sort.Slice(updater.hist[coin], func(i, j int) bool {
+		return a[i].timestamp.Before(a[j].timestamp)
+	})
+	updater.hist[coin] = a
+
+	updater.Notify(observable.Event{
+		Subject: "historical_rates",
+		Action:  action.Reload,
+	})
+	return nil
+}
+
+func (updater *RateUpdater) histLatestTimestamp(coin string) time.Time {
+	updater.histMu.RLock()
+	defer updater.histMu.RUnlock()
+	var t time.Time
+	if n := len(updater.hist[coin]); n > 0 {
+		t = updater.hist[coin][n-1].timestamp
+	}
+	return t
+}
+
+func (updater *RateUpdater) histEarliestTimestamp(coin string) time.Time {
+	updater.histMu.RLock()
+	defer updater.histMu.RUnlock()
+	var t time.Time
+	if len(updater.hist[coin]) > 0 {
+		t = updater.hist[coin][0].timestamp
+	}
+	return t
+}
+
+func (updater *RateUpdater) fetchMarketRange(ctx context.Context, coin, fiat string, start, end time.Time) ([]exchangeRate, error) {
+	param := url.Values{
+		"from":        {strconv.FormatInt(start.Unix(), 10)},
```

Comment:
should double check if it is the right resolution, same as in the response, we have to divide by `1000`, this could also be off by a factor.

though the last 3 months seems to work, so probably ok.

Reply:
yes, it expects unix, i.e. seconds.